pkg/sql/opt/optbuilder/mutation_builder.go

```diff
@@ -803,6 +803,10 @@ func (mb *mutationBuilder) addCheckConstraintCols(isUpdate bool) {
 
 		for i, n := 0, mb.tab.CheckCount(); i < n; i++ {
 			check := mb.tab.Check(i)
+			// Only enforce the check constraint if it has been validated.
+			if !check.Validated() {
+				continue
+			}
```

Comment:
This part is outside of my knowledge since it's in optimizer land. I hope someone from Queries can take a look.

In schema land, a check constraint transitions from ABSENT -> WRITE_ONLY -> VALIDATED -> PUBLIC where you can see there are two intermediate states:
  1. WRITE_ONLY (sometime you also see VALIDATING; it's the same thing): check constraint in this state means it's enforced for incoming writes 
  2. VALIDATED: check constraint in this state means it's validated on existing data
This way, we ensure that by the time the constraint is public, all data met the constraints.